File: tests/c/test_policy.c
#include <qhw_admission/qhw_admission.h>

#include <stdint.h>
#include <stdio.h>
#include <string.h>

#define CHECK(cond) \
	do { \
		if (!(cond)) { \
			fprintf(stderr, "check failed at %s:%d: %s\n", \
				__FILE__, __LINE__, #cond); \
			return 1; \
		} \
	} while (0)

static qhw_adm_baseline_t make_baseline(void)
{
	qhw_adm_baseline_t baseline = {
		.struct_size = sizeof(baseline),
		.qubit_count = 4,
		.depth = 10,
		.one_q_gate_count = 10,
		.two_q_gate_count = 5,
		.shots = 100,
		.measurement_count = 2,
	};

	return baseline;
}

static qhw_adm_device_profile_t make_profile(void)
{
	qhw_adm_device_profile_t profile = {
		.struct_size = sizeof(profile),
		.device_id = 7,
		.time_span_ns = 1000000000ULL,
		.baseline = make_baseline(),
		.max_qubits = 20,
		.max_shots = 10000,
		.one_q_gate_ns = 20,
		.two_q_gate_ns = 100,
		.measurement_ns = 1000,
		.one_q_gate_transfer_ns = 1,
		.two_q_gate_transfer_ns = 4,
		.measurement_transfer_ns = 10,
		.compile_ns = 1000,
		.control_overhead_ns = 200,
		.provider_overhead_ns = 300,
		.total_credits = 100,
		.device_rate = 50,
		.concurrent_jobs = 4,
		.default_ttl_ns = 60000000000ULL,
	};

	return profile;
}

static qhw_adm_qtask_class_t make_task(void)
{
	qhw_adm_qtask_class_t task = {
		.struct_size = sizeof(task),
		.class_id = 11,
		.count = 2,
		.qubit_count = 4,
		.depth = 12,
		.one_q_gate_count = 20,
		.two_q_gate_count = 10,
		.shots = 100,
		.measurement_count = 2,
	};

	return task;
}

static qhw_adm_request_t make_request(qhw_adm_qtask_class_t *task)
{
	qhw_adm_request_t request = {
		.struct_size = sizeof(request),
		.request_id = 42,
		.device_id = 7,
		.user_id = 1000,
		.job_id = 2000,
		.scope_id = 3,
		.workload_kind = QHW_ADM_WORKLOAD_HYBRID_JOB,
		.walltime_ns = 2000000000ULL,
		.ttl_ns = 1000000ULL,
		.task_class_count = 1,
		.task_classes = task,
	};

	return request;
}

static qhw_adm_decision_t make_decision_output(void)
{
	qhw_adm_decision_t decision;

	memset(&decision, 0, sizeof(decision));
	decision.struct_size = sizeof(decision);
	return decision;
}

static qhw_adm_capacity_view_t make_capacity_output(void)
{
	qhw_adm_capacity_view_t capacity;

	memset(&capacity, 0, sizeof(capacity));
	capacity.struct_size = sizeof(capacity);
	return capacity;
}

static qhw_adm_reservation_t make_reservation_output(void)
{
	qhw_adm_reservation_t reservation;

	memset(&reservation, 0, sizeof(reservation));
	reservation.struct_size = sizeof(reservation);
	return reservation;
}

static int setup_context(qhw_adm_t **out_ctx)
{
	qhw_adm_t *ctx = NULL;
	qhw_adm_device_profile_t profile = make_profile();

	CHECK(qhw_adm_create(NULL, &ctx) == QHW_ADM_OK);
	CHECK(qhw_adm_register_device(ctx, &profile) == QHW_ADM_OK);
	CHECK(qhw_adm_add_policy_path(ctx, QHW_ADM_TEST_POLICY_DIR) ==
		QHW_ADM_OK);
	CHECK(qhw_adm_set_policy(ctx, profile.device_id, "mock", NULL, 0) ==
		QHW_ADM_OK);
	*out_ctx = ctx;
	return 0;
}

static int test_policy_load_and_evaluate(void)
{
	qhw_adm_t *ctx = NULL;
	qhw_adm_policy_t *policy = NULL;
	qhw_adm_device_profile_t profile = make_profile();
	qhw_adm_qtask_class_t task = make_task();
	qhw_adm_request_t request = make_request(&task);
	qhw_adm_decision_t decision = make_decision_output();

	CHECK(qhw_adm_create(NULL, &ctx) == QHW_ADM_OK);
	CHECK(qhw_adm_load_policy(ctx, QHW_ADM_TEST_POLICY_PATH, &policy) ==
		QHW_ADM_OK);
	CHECK(policy != NULL);
	CHECK(qhw_adm_register_device(ctx, &profile) == QHW_ADM_OK);
	CHECK(qhw_adm_set_policy(ctx, request.device_id, "mock", NULL, 0) ==
		QHW_ADM_OK);
	CHECK(qhw_adm_evaluate(ctx, &request, &decision) == QHW_ADM_OK);
	CHECK(decision.decision == QHW_ADM_DECISION_ACCEPTED);
	CHECK(decision.request_id == request.request_id);
	CHECK(decision.estimated_total_ns == 683160);
	CHECK(decision.capacity_granted == 3);
	CHECK(decision.reservation_id == 0);
	qhw_adm_destroy(ctx);
	return 0;
}

static int test_reserve_and_release(void)
{
	qhw_adm_t *ctx = NULL;
	qhw_adm_baseline_t baseline = make_baseline();
	qhw_adm_device_profile_t profile = make_profile();
	qhw_adm_qtask_class_t task = make_task();
	qhw_adm_request_t request = make_request(&task);
	qhw_adm_decision_t decision = make_decision_output();
	qhw_adm_reservation_t reservation = make_reservation_output();
	qhw_adm_capacity_view_t capacity = make_capacity_output();

	CHECK(setup_context(&ctx) == 0);
	CHECK(qhw_adm_reserve(ctx, &request, &decision) == QHW_ADM_OK);
	CHECK(decision.decision == QHW_ADM_DECISION_ACCEPTED);
	CHECK(decision.reservation_id != 0);
	CHECK(qhw_adm_get_reservation(ctx, decision.reservation_id,
		&reservation) == QHW_ADM_OK);
	CHECK(reservation.state == QHW_ADM_RESERVATION_ACTIVE);
	CHECK(reservation.credits_reserved == 3);
	CHECK(reservation.rate_reserved == 3);
	CHECK(reservation.quantum_budget_ns == request.walltime_ns);
	CHECK(reservation.expires_at_ns > reservation.created_at_ns);
	CHECK(qhw_adm_set_policy(ctx, request.device_id, "mock", NULL, 0) ==
		QHW_ADM_ERR_STATE);
	CHECK(qhw_adm_set_estimator(ctx, request.device_id, "baseline",
		NULL, 0) == QHW_ADM_ERR_STATE);
	CHECK(qhw_adm_set_baseline(ctx, request.device_id, &baseline) ==
		QHW_ADM_ERR_STATE);
	CHECK(qhw_adm_register_device(ctx, &profile) == QHW_ADM_ERR_STATE);
	CHECK(qhw_adm_unregister_device(ctx, request.device_id) ==
		QHW_ADM_ERR_STATE);

	CHECK(qhw_adm_get_capacity(ctx, request.device_id, request.scope_id,
		&capacity) == QHW_ADM_OK);
	CHECK(capacity.credits_reserved == 3);
	CHECK(capacity.scoped_reserved_credits == 3);
	CHECK(capacity.scheduler_policy_id == 1234);

	CHECK(qhw_adm_release(ctx, decision.reservation_id,
		QHW_ADM_REASON_NONE) == QHW_ADM_OK);
	reservation = make_reservation_output();
	CHECK(qhw_adm_get_reservation(ctx, decision.reservation_id,
		&reservation) == QHW_ADM_OK);
	CHECK(reservation.state == QHW_ADM_RESERVATION_RELEASED);
	CHECK(qhw_adm_release(ctx, decision.reservation_id,
		QHW_ADM_REASON_NONE) == QHW_ADM_OK);
	qhw_adm_destroy(ctx);
	return 0;
}

static int test_cancel_renew_and_expire(void)
{
	qhw_adm_t *ctx = NULL;
	qhw_adm_qtask_class_t task = make_task();
	qhw_adm_request_t request = make_request(&task);
	qhw_adm_decision_t decision = make_decision_output();
	qhw_adm_reservation_t reservation = make_reservation_output();
	size_t expired_count = 0;

	CHECK(setup_context(&ctx) == 0);
	CHECK(qhw_adm_reserve(ctx, &request, &decision) == QHW_ADM_OK);
	CHECK(qhw_adm_renew(ctx, decision.reservation_id, 1000, 1000) ==
		QHW_ADM_OK);
	CHECK(qhw_adm_expire(ctx, 1999, &expired_count) == QHW_ADM_OK);
	CHECK(expired_count == 0);
	CHECK(qhw_adm_expire(ctx, 2000, &expired_count) == QHW_ADM_OK);
	CHECK(expired_count == 1);
	CHECK(qhw_adm_get_reservation(ctx, decision.reservation_id,
		&reservation) == QHW_ADM_OK);
	CHECK(reservation.state == QHW_ADM_RESERVATION_EXPIRED);
	CHECK(qhw_adm_cancel(ctx, decision.reservation_id,
		QHW_ADM_REASON_NONE) == QHW_ADM_ERR_STATE);
	qhw_adm_destroy(ctx);

	CHECK(setup_context(&ctx) == 0);
	decision = make_decision_output();
	CHECK(qhw_adm_reserve(ctx, &request, &decision) == QHW_ADM_OK);
	CHECK(qhw_adm_cancel(ctx, decision.reservation_id,
		QHW_ADM_REASON_NONE) == QHW_ADM_OK);
	CHECK(qhw_adm_cancel(ctx, decision.reservation_id,
		QHW_ADM_REASON_NONE) == QHW_ADM_OK);
	qhw_adm_destroy(ctx);
	return 0;
}

int main(void)
{
	CHECK(test_policy_load_and_evaluate() == 0);
	CHECK(test_reserve_and_release() == 0);
	CHECK(test_cancel_renew_and_expire() == 0);
	return 0;
}

File: tests/python/test_policy.py
import os
import unittest

from qhw_admission import (
    AdmissionContext,
    AdmissionError,
    AdmissionRequest,
    Baseline,
    COMPLIANCE_ALLOW,
    DECISION_ACCEPTED,
    DeviceProfile,
    QtaskClass,
    RESERVATION_ACTIVE,
    RESERVATION_EXPIRED,
    RESERVATION_RELEASED,
    VALUE_STRING,
    WORKLOAD_HYBRID_JOB,
)


def make_profile():
    baseline = Baseline(
        qubit_count=4,
        depth=10,
        one_q_gate_count=10,
        two_q_gate_count=5,
        shots=100,
        measurement_count=2,
    )
    return DeviceProfile(
        device_id=7,
        time_span_ns=1_000_000_000,
        baseline=baseline,
        max_qubits=20,
        max_shots=10_000,
        one_q_gate_ns=20,
        two_q_gate_ns=100,
        measurement_ns=1000,
        one_q_gate_transfer_ns=1,
        two_q_gate_transfer_ns=4,
        measurement_transfer_ns=10,
        compile_ns=1000,
        control_overhead_ns=200,
        provider_overhead_ns=300,
        total_credits=100,
        device_rate=50,
        concurrent_jobs=4,
        default_ttl_ns=60_000_000_000,
    )


def make_task():
    return QtaskClass(
        class_id=11,
        count=2,
        qubit_count=4,
        depth=12,
        one_q_gate_count=20,
        two_q_gate_count=10,
        shots=100,
        measurement_count=2,
    )


def make_request(task, ttl_ns=1_000_000):
    return AdmissionRequest(
        request_id=42,
        device_id=7,
        user_id=1000,
        job_id=2000,
        scope_id=3,
        workload_kind=WORKLOAD_HYBRID_JOB,
        walltime_ns=2_000_000_000,
        ttl_ns=ttl_ns,
        task_class=task,
    )


class PolicyTests(unittest.TestCase):
    def setup_context(self):
        ctx = AdmissionContext()
        ctx.register_device(make_profile())
        ctx.add_policy_path(os.environ["QHW_ADM_TEST_POLICY_DIR"])
        ctx.set_policy(7, "mock")
        return ctx

    def test_evaluate_and_reserve(self):
        task = make_task()
        request = make_request(task)

        with self.setup_context() as ctx:
            decision = ctx.evaluate(request)
            self.assertEqual(decision.decision, DECISION_ACCEPTED)
            self.assertEqual(decision.estimated_total_ns, 683160)
            self.assertEqual(decision.capacity_granted, 3)
            self.assertEqual(decision.reservation_id, 0)
            self.assertEqual(decision.compliance_action, COMPLIANCE_ALLOW)
            self.assertEqual(decision.retry_after_ns, 0)
            self.assertEqual(decision.message, "mock accepted")
            self.assertEqual(
                decision.metadata,
                ((1001, VALUE_STRING, "decision-metadata"),),
            )

            decision = ctx.reserve(request)
            self.assertEqual(decision.decision, DECISION_ACCEPTED)
            self.assertNotEqual(decision.reservation_id, 0)
            self.assertEqual(decision.message, "mock accepted")

            reservation = ctx.get_reservation(decision.reservation_id)
            self.assertEqual(reservation.state, RESERVATION_ACTIVE)
            self.assertEqual(reservation.credits_reserved, 3)
            self.assertEqual(reservation.credits_consumed, 0)
            self.assertEqual(reservation.rate_reserved, 3)
            self.assertEqual(reservation.rate_consumed, 0)
            self.assertEqual(reservation.actual_total_ns, 0)
            self.assertEqual(reservation.unused_capacity, 0)
            self.assertEqual(reservation.overuse_count, 0)
            self.assertEqual(reservation.underuse_score, 0)
            self.assertGreater(reservation.device_profile_version, 0)
            self.assertGreater(reservation.policy_version, 0)
            self.assertGreater(reservation.estimator_version, 0)
            self.assertEqual(
                reservation.metadata,
                ((1002, VALUE_STRING, "grant-metadata"),),
            )

            capacity = ctx.get_capacity(7, 3)
            self.assertEqual(capacity.credits_reserved, 3)
            self.assertEqual(capacity.credits_consumed, 0)
            self.assertEqual(capacity.credits_returned, 0)
            self.assertEqual(capacity.scoped_reserved_credits, 3)
            self.assertEqual(capacity.scheduler_policy_id, 1234)
            self.assertEqual(capacity.confidence_ppm, 1_000_000)
            self.assertEqual(
                capacity.metadata,
                ((1003, VALUE_STRING, "capacity-metadata"),),
            )

            with self.assertRaises(AdmissionError):
                ctx.set_policy(7, "mock")

            ctx.release(decision.reservation_id)
            reservation = ctx.get_reservation(decision.reservation_id)
            self.assertEqual(reservation.state, RESERVATION_RELEASED)

    def test_renew_and_expire(self):
        task = make_task()
        request = make_request(task)

        with self.setup_context() as ctx:
            decision = ctx.reserve(request)
            ctx.renew(decision.reservation_id, 1000, 1000)
            self.assertEqual(ctx.expire(1999), 0)
            self.assertEqual(ctx.expire(2000), 1)
            reservation = ctx.get_reservation(decision.reservation_id)
            self.assertEqual(reservation.state, RESERVATION_EXPIRED)


if __name__ == "__main__":
    unittest.main()
